### DISCUSSION TITLE
Serialising back to XML

### DISCUSSION THREAD

**Discussion by jaystey** [rgrove/parse-xml#21] [Ideas]
Hi there,

I'm wondering if you would consider adding functionality such that XmlDocument (and perhaps XmlNode) objects could be serialised back to strings of XML.

Thanks for a great library!

**Comment by rgrove** [comment]
Thanks! I'm glad you like the library.

I think it would be interesting to write a serializer, but it's probably not something I'm going to have time to tackle soon. If I did, though, I would implement it as a separate library (such as `@rgrove/serialize-xml`). Thanks for letting me know you'd find that useful.

**Comment by jahow** [comment]
For anyone interested, serializing an XmlDocument back to a string is actually pretty straightforward. Here is a serializing function which contains a few deliberate formatting choices and ignores some nodes (comments, CDATA...):

```ts
export function xmlToString(
  el: XmlElement | XmlText | XmlComment | XmlDocument,
  indentationLevel = 0
) {
  if (el.constructor.name === 'XmlDocument')
    return `<?xml version="1.0" encoding="UTF-8"?>${xmlToString(
      (el as XmlDocument).children[0]
    )}`
  if (el.constructor.name === 'XmlText') {
    const text = (el as XmlText).text
    const isEmpty = text.replace(/^\s+|\s+$/g, '') === ''
    if (isEmpty) return ''
    return text
  }
  if (el.constructor.name !== 'XmlElement' || !(el instanceof XmlElement))
    return `<!-- unknown -->`

  const padding = '    '.repeat(indentationLevel)
  const children = Array.isArray(el.children)
    ? el.children
        .map((el) => xmlToString(el, indentationLevel + 1))
        .filter((el) => el !== '')
        .map((elString, index, array) =>
          index < array.length - 1 ? elString.replace(/\n\s*$/g, '') : elString
        )
        .join('')
    : ''
  const attrs = Object.keys(el.attributes).reduce(
    (prev, curr) => prev + ` ${curr}="${el.attributes[curr]}"`,
    ''
  )
  const parentPadding = '    '.repeat(Math.max(0, indentationLevel - 1))
  if (children === '') {
    return `
${padding}<${el.name}${attrs}/>
${parentPadding}`
  }
  return `
${padding}<${el.name}${attrs}>${children}</${el.name}>
${parentPadding}`
}
```

The function is recursive so you just have to call it on the root XmlDocument to have everything printed out.

**Reply by josephjunker** [reply]
Hi @jahow , thank you for providing this sample. This is extremely useful. Would you mind specifying whether you are releasing this code under a license (hopefully MIT or CC0)? I'd like to use it inside of a closed-source project but am not comfortable copying this amount of unlicensed code.

**Reply by jahow** [reply]
hey! since this snippet was shared in the parse-xml repository which is in ISD (same as MIT) I guess it would make sense to consider it identically licensed.

Also thanks a lot for having that respectful approach! :pray:

**Reply by josephjunker** [reply]
Great, thanks a ton!
